libfolly: Add package for folly library

ammubhave opened the pull request:
Depends on ~PR #7098~, ~PR #7126~, ~PR #7877~

Maintainer: me
Compile tested: openwrt-18.06 for ipq40xx, ipq806x, x86 and ar71xx
Run tested: Tested on devices running on the above architectures. Verified that dependent packages can successfully build and run using the libfolly.so shared library

Description: folly is an open-source C++ library developed and used at Facebook (https://github.com/facebook/folly).
Adding this package will enable adding further packages that depend on this.

diizzyy commented on libs/libfolly/Makefile:
```diff
@@ -0,0 +1,49 @@
+endef
+
+define Package/libfolly/description
+	An open-source C++ library developed and used at Facebook.
```
Please make it more informative

diizzyy commented on libs/libfolly/Makefile:
```diff
@@ -0,0 +1,49 @@
+	-DHAVE_VSNPRINTF_ERRORS_EXITCODE=OFF \
+	-DFOLLY_HAVE_XSI_STRERROR_R_EXITCODE=0 \
+	-DBUILD_SHARED_LIBS=ON
+EXTRA_CXXFLAGS:=-Wno-cpp -Wno-maybe-uninitialized -fPIC
```
```$(FPIC)```?

diizzyy commented on libs/libfolly/Makefile:
```diff
@@ -0,0 +1,49 @@
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://github.com/facebook/folly.git
```
Use codeload and import the latest version

diizzyy commented:
Changes and adding packages (glog, libgflags, libmstch, zstd) should be in separate PRs

neheb commented on libs/libfolly/Makefile:
```diff
@@ -0,0 +1,79 @@
+	SECTION:=libs
+	CATEGORY:=Libraries
+	TITLE:=An open-source C++ library developed and used at Facebook.
+  URL:=https://github.com/facebook/folly
```
Mismatching whitespace?

neheb commented:
This is failing on uClibc-ng:

```
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/portability/Math.h:29:33: error: 'std::nextafter' has not been declared
 /* using override */ using std::nextafter;
                                 ^~~~~~~~~
In file included from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/Conv.cpp:16:
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/Conv.h: In function 'typename std::enable_if<((std::is_floating_point<Src>::value && std::is_integral<_Tp>::value) && (! std::is_same<T, bool>::value)), bool>::type folly::detail::checkConversion(const Src&)':
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/Conv.h:1274:29: error: 'nextafter' is not a member of 'folly'
     const Src mmax = folly::nextafter(tgtMaxAsSrc, Src());
                             ^~~~~~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/Conv.h:1274:29: note: suggested alternative: 'unexpected'
     const Src mmax = folly::nextafter(tgtMaxAsSrc, Src());
                             ^~~~~~~~~
                             unexpected
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/Conv.h:1283:29: error: 'nextafter' is not a member of 'folly'
     const Src mmin = folly::nextafter(tgtMinAsSrc, Src());
                             ^~~~~~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/Conv.h:1283:29: note: suggested alternative: 'unexpected'
     const Src mmin = folly::nextafter(tgtMinAsSrc, Src());
                             ^~~~~~~~~
                             unexpected
CMakeFiles/folly_base.dir/build.make:88: recipe for target 'CMakeFiles/folly_base.dir/folly/Conv.cpp.o' failed
```

The reason for this is that nextafter is defined in tgmath.h only, not in math.h as in musl.

neheb commented:
Another one:

```
In file included from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/synchronization/LifoSem.h:26,
                 from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/executors/task_queue/LifoSemMPMCQueue.h:21,
                 from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/executors/ThreadPoolExecutor.h:26,
                 from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/executors/CPUThreadPoolExecutor.h:19,
                 from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/executors/CPUThreadPoolExecutor.cpp:17:
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/CachelinePadded.h: In instantiation of 'class folly::CachelinePadded<folly::AtomicStruct<folly::detail::LifoSemHead, std::atomic> >':
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/synchronization/LifoSem.h:591:59:   required from 'struct folly::detail::LifoSemBase<folly::SaturatingSemaphore<true, std::atomic>, std::atomic>'
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/synchronization/LifoSem.h:732:8:   required from 'struct folly::LifoSemImpl<>'
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/executors/task_queue/LifoSemMPMCQueue.h:72:18:   required from here
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/CachelinePadded.h:39:18: error: static assertion failed: CachelinePadded does not support over-aligned types.
       alignof(T) <= max_align_v,
       ~~~~~~~~~~~^~~~~~~~~~~~~~
CMakeFiles/folly_base.dir/build.make:712: recipe for target 'CMakeFiles/folly_base.dir/folly/executors/CPUThreadPoolExecutor.cpp.o' failed
```

ammubhave commented:
Added patches to make it work using uclibc as well.

- CachelinePadded static_assert is a stronger assertion than it needs to be, it can be safely removed.
- Added a few of patches to work around unavailability of std::remainder and std::nextafter. I'll create diffs for upstream for most of these patches as well.

neheb commented on libs/libfolly/Makefile:
```diff
@@ -0,0 +1,55 @@
+	-DHAVE_VSNPRINTF_ERRORS_EXITCODE=OFF \
+	-DFOLLY_HAVE_XSI_STRERROR_R_EXITCODE=0 \
+	-DBUILD_SHARED_LIBS=ON
+EXTRA_CXXFLAGS:=-Wno-cpp -Wno-maybe-uninitialized -Wno-attributes -Wno-terminate
```
Are these necessary? Compilation locally succeeded without these.

ammubhave replied:
You're right, this release no longer requires them.

neheb commented on libs/libfolly/Makefile:
```diff
@@ -0,0 +1,55 @@
+
+PKG_BUILD_PARALLEL:=1
+CMAKE_OPTIONS:= \
+	-DCMAKE_POSITION_INDEPENDENT_CODE=True \
```
ON would be more consistent here.

neheb commented:
Different error this time:

```
[ 52%] Building CXX object CMakeFiles/folly_base.dir/folly/io/async/AsyncPipe.cpp.o
In file included from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/io/async/AsyncTransportCertificate.h:20,
                 from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/io/async/AsyncTransport.h:23,
                 from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/io/async/AsyncPipe.h:23,
                 from /home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/io/async/AsyncPipe.cpp:16:
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:31:73: error: template argument 2 is invalid
   using alias##Deleter = folly::static_function_deleter<object, &deleter>; \
                                                                         ^
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:108:1: note: in expansion of macro 'DEFINE_SSL_PTR_TYPE'
 DEFINE_SSL_PTR_TYPE(BIGNUM, BIGNUM, BN_clear_free);
 ^~~~~~~~~~~~~~~~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:108:21: error: 'BIGNUMDeleter' was not declared in this scope
 DEFINE_SSL_PTR_TYPE(BIGNUM, BIGNUM, BN_clear_free);
                     ^~~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:32:52: note: in definition of macro 'DEFINE_SSL_PTR_TYPE'
   using alias##UniquePtr = std::unique_ptr<object, alias##Deleter>
                                                    ^~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:108:21: note: suggested alternative: 'BioDeleter'
 DEFINE_SSL_PTR_TYPE(BIGNUM, BIGNUM, BN_clear_free);
                     ^~~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:32:52: note: in definition of macro 'DEFINE_SSL_PTR_TYPE'
   using alias##UniquePtr = std::unique_ptr<object, alias##Deleter>
                                                    ^~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:32:66: error: template argument 2 is invalid
   using alias##UniquePtr = std::unique_ptr<object, alias##Deleter>
                                                                  ^
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:108:1: note: in expansion of macro 'DEFINE_SSL_PTR_TYPE'
 DEFINE_SSL_PTR_TYPE(BIGNUM, BIGNUM, BN_clear_free);
 ^~~~~~~~~~~~~~~~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:31:73: error: template argument 2 is invalid
   using alias##Deleter = folly::static_function_deleter<object, &deleter>; \
                                                                         ^
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:109:1: note: in expansion of macro 'DEFINE_SSL_PTR_TYPE'
 DEFINE_SSL_PTR_TYPE(BNCtx, BN_CTX, BN_CTX_free);
 ^~~~~~~~~~~~~~~~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:109:21: error: 'BNCtxDeleter' was not declared in this scope
 DEFINE_SSL_PTR_TYPE(BNCtx, BN_CTX, BN_CTX_free);
                     ^~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:32:52: note: in definition of macro 'DEFINE_SSL_PTR_TYPE'
   using alias##UniquePtr = std::unique_ptr<object, alias##Deleter>
                                                    ^~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:109:21: note: suggested alternative: 'BioDeleter'
 DEFINE_SSL_PTR_TYPE(BNCtx, BN_CTX, BN_CTX_free);
                     ^~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:32:52: note: in definition of macro 'DEFINE_SSL_PTR_TYPE'
   using alias##UniquePtr = std::unique_ptr<object, alias##Deleter>
                                                    ^~~~~
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:32:66: error: template argument 2 is invalid
   using alias##UniquePtr = std::unique_ptr<object, alias##Deleter>
                                                                  ^
/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00/folly/ssl/OpenSSLPtrTypes.h:109:1: note: in expansion of macro 'DEFINE_SSL_PTR_TYPE'
 DEFINE_SSL_PTR_TYPE(BNCtx, BN_CTX, BN_CTX_free);
 ^~~~~~~~~~~~~~~~~~~
CMakeFiles/folly_base.dir/build.make:1635: recipe for target 'CMakeFiles/folly_base.dir/folly/io/async/AsyncPipe.cpp.o' failed
make[6]: *** [CMakeFiles/folly_base.dir/folly/io/async/AsyncPipe.cpp.o] Error 1
make[6]: Leaving directory '/home/mangix/devstuff/openwrt/build_dir/target-arc_arc700_uClibc/folly-2019.05.06.00'
CMakeFiles/Makefile2:146: recipe for target 'CMakeFiles/folly_base.dir/all' failed
make[5]: *** [CMakeFiles/folly_base.dir/all] Error 2
```

This one is probably caused by disabling deprecated OpenSSL APIs (found in Libraries > libssl). Based on these, it seems the issue is a missing include to openssl/bn.h

ammubhave commented:
Yeah, I repro'd it by disabling the deprecated APIs as well. folly's ssl wrappers depend on deprecated api for now. (I added openssl/bn.h, but it seems there are several other places where the deprecated api is used).
I added a DEPENDS on @OPENSSL_WITH_DEPRECATED

ammubhave commented:
Actually on that note, do you know of any way of making it such that whenever libfolly gets selected, CONFIG_ OPENSSL_WITH_DEPRECATED automatically gets selected?

neheb commented:
Add a + in front of the @ .

ammubhave commented:
oh lol, that was easy. I previously thought that only worked for selected packages.

neheb commented:
I'm getting this:

```
Package libfolly is missing dependencies for the following libraries:
libaio.so.1
```

What probably happened is I selected libaio and then deselected it, making folly pick up the library.

From a quick look, this could also be a problem with SNAPPY and LZ4. These need to be explicitly enabled or disabled.

As for the deprecated APIs, I wrote a patch:

https://gist.github.com/neheb/997e33ae81107f7ebccc3b032189d02c

No runtime testing of any kind but it does compile.

ammubhave commented:
folly picked up libaio. looks like it is an optional dependency (https://github.com/facebook/folly/blob/master/CMakeLists.txt#L178). No harm in always enabling it.
I ran the unit tests under folly/ssl/test with your patch with openssl 1.0.2 and they passed, so you can upstream it if that's preferable. I included it in the patches/ for now and removed the deprecated_api config dependency.

neheb commented:
Yeah, buildbots are throwing this:

```
Package libfolly is missing dependencies for the following libraries:
liblz4.so.1
libunwind.so.8
```
